Restructuring of deepquest for packaging

davidwyld opened the pull request:
This PR incorporates a major restructuring of the project to:
1. Make it a pip installable package.
2. Remove unnecessary components of the codebase, mainly NMT-Keras, to use the master repo of that project (lvapeab/nmt-keras)
    - This included making PRs to lvapeab/nmt-keras in order to package that project to be usable by deepquest.
3. Move `dq_tests` to `tests` and `dq_utils` to `deepquest/utils`.
4. Move the models directory to `deepquest/qe_models`
5. Collate all data preparation functions into `deepquest/utils/prepare_data.py` including `preprocessDoc`.

Closes https://github.com/RSE-Sheffield/deepquest/issues/28 and closes https://github.com/RSE-Sheffield/deepquest/issues/3

At present, in order to run deepquest, it must be installed as a package either by:
1. Cloning the repo and running `pip install .` for `setup.py`, or
2. Installing from the repo directly as `pip install 'deepquest @ https://github.com/rse-sheffield/nmt-keras/archive/master.zip'`
    - Note that this option will not copy all files from the repo, but only those required for running the package.

deepquest can be run from any directory, after installation, either from the command line as one of the following:
```shell
deepquest train -c configs/config.yml
dq train -c configs/config.yml
```
or via importing as a python module, for instance:
```python
import deepquest as dq
dq.train('configs/config.yml')
```

Currently, deepquest will look in the present working directory for all data and config files, and put the `trained_models` directory there also.

**Note**: as a python module, seed setting is not implemented, as this must happen before importing any of the modules that use the seed. So, for deterministic results, one would need to do the following before importing deepquest.
```python
seed = 1
import numpy.random
numpy.random.seed(seed)
import random
random.seed(seed)
```

Commit:
delete old config file

Commit:
set relative imports

Commit:
add __init__ files

Commit:
improve cmd line documentation

Commit:
help and error handling

Commit:
modify nmt-keras relative path

Commit:
correct arg to config

Commit:
remove imports

Commit:
rename nmt_keras to quest

Commit:
remove utils

Commit:
update imports

Commit:
remove relative imports

Commit:
use deepquest's prepare_data

Commit:
temporarily point requirements at davidwilby/nmt-keras

Commit:
use lvapeab/nmt-keras again

Commit:
remove unused old nmt-keras files

Commit:
update requirements to use PyPI multimodal-keras-wrapper

Commit:
rename 'dq_tests' directory to 'tests'

Commit:
rename 'dq_utils' directory to 'utils'

Commit:
move dq_evaluation to utils/evaluation.py

Commit:
remove data_engine and use utils/prepare_data

Commit:
move preprocessDoc to utils/prepare_data

Commit:
move all callbacks into utils/callbacks.py

Commit:
move quest/models to qe_models

Commit:
rearrange imports

Commit:
fix all imports

Commit:
add dq.py

Commit:
restructure the package

Commit:
make alterations to .travis.yml

Commit:
add MarcBS/keras to setup.py

drj11 commented on deepquest/predict.py:
```diff
@@ -247,7 +247,7 @@ def main(model, dataset, save_path=None, evalset=None, changes={}):
     # from nmt_keras import model_zoo
     from keras.utils import CustomObjectScope
-    import nmt_keras.models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
+    import deepquest.qe_models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
```
comment needs updating to reflect where layers come from

drj11 commented on deepquest/train.py:
```diff
@@ -138,7 +138,7 @@ def train_model(params, weights_dict, load_dataset=None, trainable_pred=True, tr
             # from the files containing the model
             from keras.utils import CustomObjectScope
-            import nmt_keras.models.utils as layers  # includes all layers and everything defined in nmt_keras.utils
+            from deepquest.qe_models import utils as layers  # includes all layers and everything defined in nmt_keras.utils
```
same as for predict, comment needs updating.

drj11 commented on deepquest/train.py:
```diff
@@ -320,21 +320,21 @@ def main(config=None, changes={}):
-        if arg.endswith('.yml'):
+        if config.endswith('.yml'):
             # FIXME make this a user option (maybe depend on model type and level?)
             with open('configs/default-config-BiRNN.yml') as file:
```
I'll come over and ask you about this.

drj11 commented on deepquest/utils/prepare_data.py:
```diff
@@ -9,6 +9,74 @@
 # We patch the Dataset class here.
 # This will not be necessary when multimodal_keras_wrapper is fixed.
```
This comment is supposed to be attached to, and precede, the `dataset_preprocessreal_fixed` function. But you slapped a gigantic new function in between this comment and the function that follows. I suppose that's my fault for putting blank lines between them.

davidwyld replied:
Quite right! Have actually just removed `dataset_preprocessreal_fixed` in a6c9376 since this is now fixed in multimodal_keras_wrapper.

drj11 commented on setup.py:
```diff
@@ -0,0 +1,41 @@
+    ],
+    install_requires=[
+        "future",
+        "numpy<1.17,>=1.14.5",
```
that's slightly surprising and might need a comment to justify it

davidwyld replied:
Note to self: is this not set in the TensorFlow requirements?

davidwyld replied:
There was an incompatibility warning between tensorflow and numpy, but this has now been resolved as discussed here https://github.com/tensorflow/tensorflow/pull/30559#issuecomment-518258441
Have removed these requirements here.

drj11 reviewed (commented):
cool!

Commit:
Remove configs/__init__.py

This was the wrong way to include this dir in the package.
setup.py now includes this directory using package_data.

Commit:
update comment and imports for layer objects

Commit:
Remove dataset_preprocessreal_fixed temporary fix for MKW.

DeepQuest had a temporary patchfix to solve a problem in multimodal_keras_wrapper
which has now been fixed, so this function and patch is no longer needed.

Commit:
Relax numpy and tensorflow requirements

A previous numpy and tensorflow version incompatibility has been resolved.
